src/lib/player/AudioPlayer.ts

```diff
@@ -96,87 +93,111 @@ class AudioPlayer {
+            if (song) {
+                this.currentSong = song;
+            }
+        });
 
-            if (playlist.length === 0) {
+        queue.subscribe(async (queue) => {
+            if (queue.length === 0) {
+                this.queue = queue;
+                this.currentSongIdx = -1;
                 return;
             }
 
+            if (!this.currentSong) {
+                this.currentSong = get(current).song;
+            }
+
             const shuffleEnabled = get(isShuffleEnabled);
-            const shuffledPlylist = get(shuffledPlaylist);
             let newCurrentSongIdx = 0;
-            console.log("shuffled", shuffledPlylist);
+
             // If shuffle is enabled but not yet shuffled
             if (shuffleEnabled) {
-                this.shuffle();
-                // If shuffle is disabled, need to unshuffle
-            } else if (!shuffleEnabled && shuffledPlylist.length > 0) {
+                if (this.shouldShuffle) {
+                    this.shuffle();
+                } else {
+                    this.queue = get(shuffledQueue);
+
+                    const id = this.currentSong?.id;
+                    if (id) {
+                        newCurrentSongIdx = this.queue.findIndex((s) => s.id === id);
+                    } else {
+                        newCurrentSongIdx = 0;
+                    }
+
+                    this.shouldShuffle = true;
+                }
+            // If shuffle is disabled, need to unshuffle
+            } else if (!shuffleEnabled && get(shuffledQueue).length > 0) {
                 this.unshuffle();
                 // Restore position
-                newCurrentSongIdx = playlist.findIndex(
-                    (s) => s.id === this.currentSong?.id
-                );
-                if (newCurrentSongIdx === -1) {
+                const id = this.currentSong?.id;
+                if (id) {
+                    newCurrentSongIdx = this.queue.findIndex((s) => s.id === id);
+                } else {
                     newCurrentSongIdx = 0;
                 }
+            // Shuffle was disabled already
             } else {
-                // Shuffle was disabled already
-                this.playlist = playlist;
+                this.queue = queue;
 
-                newCurrentSongIdx = playlist.findIndex(
-                    (s) => s.id === this.currentSong?.id
-                );
-                // console.log("found index", playlist, this.currentSong);
-                if (newCurrentSongIdx === -1) {
+                const id = this.currentSong?.id;
+                if (id) {
+                    newCurrentSongIdx = this.queue.findIndex((s) => s.id === id);
+                } else {
                     newCurrentSongIdx = 0;
                 }
             }
 
             this.currentSongIdx = newCurrentSongIdx;
-            currentSongIdx.set(newCurrentSongIdx);
+
+            current.set({ ...get(current), index: newCurrentSongIdx });
```

Comment:
This would also unnecessarily trigger the `current` subscriber in the Sidebar though (and any usage of `$current`, refreshing the same song + artwork even though it doesn't care about the new index. Watch out for these side-effects, the reason why I had the index as a separate writable was to only update what's required. Admittedly I should have documented it better..

Reply:
Ah ok, yeah a song check will also do the trick. In the sidebar the song artwork takes priority over the album one I think, so it could happen that two songs in the same album show different artworks.

Reply:
I've made the changes